coderd/notifications/notificationstest/fake_enqueuer.go

```diff
@@ -0,0 +1,100 @@
+package notificationstest
+
+import (
+	"context"
+	"sync"
+
+	"github.com/google/uuid"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/coder/coder/v2/coderd/database/dbauthz"
+	"github.com/coder/coder/v2/coderd/rbac"
+	"github.com/coder/coder/v2/coderd/rbac/policy"
+)
+
+type FakeEnqueuer struct {
+	authorizer rbac.Authorizer
+	mu         sync.Mutex
+	sent       []*FakeNotification
+}
+
+type FakeNotification struct {
+	UserID, TemplateID uuid.UUID
+	Labels             map[string]string
+	Data               map[string]any
+	CreatedBy          string
+	Targets            []uuid.UUID
+}
+
+func (f *FakeEnqueuer) assertRBACNoLock(ctx context.Context) {
+	if f.mu.TryLock() {
+		panic("Developer error: do not call assertRBACNoLock outside of a mutex lock!")
+	}
+
+	// If we get here, we are locked.
+	if f.authorizer == nil {
+		f.authorizer = rbac.NewStrictCachingAuthorizer(prometheus.NewRegistry())
+	}
+
+	act, ok := dbauthz.ActorFromContext(ctx)
+	if !ok {
+		panic("Developer error: no actor in context, you may need to use dbauthz.AsNotifier(ctx)")
+	}
+
+	err := f.authorizer.Authorize(ctx, act, policy.ActionCreate, rbac.ResourceNotificationMessage)
+	if err == nil {
+		return
+	}
+
+	if rbac.IsUnauthorizedError(err) {
+		panic("Developer error: not authorized to send notification msg. " +
+			"Ensure that you are using dbauthz.AsXXX with an actor that has " +
+			"policy.ActionCreate on rbac.ResourceNotificationMessage")
+	}
+	panic("Developer error: failed to check auth:" + err.Error())
+}
+
+func (f *FakeEnqueuer) Enqueue(ctx context.Context, userID, templateID uuid.UUID, labels map[string]string, createdBy string, targets ...uuid.UUID) (*uuid.UUID, error) {
+	f.mu.Lock()
```

Comment:
nit: should we call `Enqueue` instead?

Reply:
The original implementation called `EnqueueWithData`:

```
func (f *FakeNotificationsEnqueuer) Enqueue(ctx context.Context, userID, templateID uuid.UUID, labels map[string]string, createdBy string, targets ...uuid.UUID) (*uuid.UUID, error) {
	return f.EnqueueWithData(ctx, userID, templateID, labels, nil, createdBy, targets...)
}
```

Reply:
sorry `EnqueueWithData` of course, but question stays the same

Reply:
Ah I see what you mean. Right now we do `Enqueue` -> `enqueueWithDataNoLock`. You're suggesting we do `Enqueue` -> `EnqueueWithData` -> lock -> `enqueueWithDataLock` or similar?

Reply:
Done as suggested